lib/rules/valid-typeof.js

```diff
@@ -74,6 +96,20 @@ module.exports = {
                             }
                         } else if (requireStringLiterals && !isTypeofExpression(sibling)) {
                             context.report({ node: sibling, messageId: "notString" });
+                        } else if (sibling.type === "Identifier" && sibling.name === "undefined" && isReferenceToGlobalVariable(sibling)) {
+                            context.report({
+                                node: sibling,
+                                messageId: "invalidValue",
+                                data: { type: "undefined" },
+                                suggest: [
+                                    {
+                                        messageId: "suggestString",
+                                        fix(fixer) {
+                                            return fixer.replaceText(sibling, '"undefined"');
+                                        }
+                                    }
+                                ]
```

Comment:
The suggestion to replace `undefined` with `"undefined"` makes sense with `requireStringLiterals: true`, too.

Maybe we could switch the order of `else-if`s:

```js
} else if (sibling.type === "Identifier" && sibling.name === "undefined" && isReferenceToGlobalVariable(sibling)) {
    context.report({
        node: sibling,
        messageId: requireStringLiterals ? "notString" : "invalidValue",
        suggest: [
            {
                messageId: "suggestString",
                data: { type: "undefined" },
                fix(fixer) {
                    return fixer.replaceText(sibling, '"undefined"');
                }
            }
        ]
    });
} else if (requireStringLiterals && !isTypeofExpression(sibling)) {
    context.report({ node: sibling, messageId: "notString" });
}
```

Test case:

```js
{
    code: "if (typeof bar !== undefined) {}",
    options: [{ requireStringLiterals: true }],
    errors: [
        {
            messageId: "notString",
            type: "Identifier",
            suggestions: [
                {
                    messageId: "suggestString",
                    data: { type: "undefined" },
                    output: 'if (typeof bar !== "undefined") {}'
                }
            ]
        }
    ]
},
```